[extension/opampextension]: Adding AcceptsRestartCommand capability (#45056)
open-telemetry/opentelemetry-collector-contrib: extension/opampextension/README.md

  - `reports_effective_config`: Whether to enable the OpAMP ReportsEffectiveConfig capability. Default is `true`.
  - `reports_health`: Whether to enable the OpAMP ReportsHealth capability. Default is `true`.
  - `reports_available_components`: Whether to enable the OpAMP ReportsAvailableComponents capability. Default is `true`.

>   - `accepts_restart_command`: Whether to enable the OpAMP AcceptsRestartCommand capability. Default is `false`. The extension sends a `SIGHUP` signal to the collector to initiate a restart (however, SIGHUP isn't supported on windows systems, so it will be ignored). This functionality is also behind a feature gate (alpha) called `extension.opampextension.RemoteRestarts`

Reviewer: Can we add a separate section in the readme that describes the following:

1. When you would want to use this functionality, and how it could be used.
2. A little more on what the SIGHUP does.
3. Bold (and possibly underlined) text stating that invalid config will cause the Collector to crash and stay down until it is restarted using some other mechanism (the extension will be down and the restart command will be unavailable). This is the riskiest part of this feature, in my opinion. We should call out that regular healthchecks and using the `validate` subcommand on the Collector are highly recommended when using this feature to help guard against data loss due to extended downtime.

See the [opampcustommessages](../opampcustommessages/README.md) module for more information on the custom message API.

## Using the AcceptsRestartCommand capability to orchestrate collector config updates
If the `accepts_restart_command` capability is enabled (along with the `extension.opampextension.RemoteRestarts` feature gate), upon receiving a restart `ServerToAgentCommand`, the extension will send a `SIGHUP` signal to the collector process. 

The `SIGHUP` signal is already trapped in the collector and initiates a graceful restart of all the components. This functionality might be desired if the collectors config was updated and the restart would fetch the latest config values.

**NOTE** on the `SIGHUP` signal for restarts:
SIGHUP is not supported in the `windows` operating system, and the current implementation will short circuit evaluation on windows systems (until the collector can also handle a cross-platform signal like SIGUSR2).

> **<ins>Please note that an invalid config (capability enabled without feature gate)</ins>** will cause the collector to error out and stay down until it is manually restarted using some other mechanism (since the opamp extension won't be active). Regular healthchecks and using the `validate` subcommand on the collector are **highly recommended** when using this functionality to prevent data loss due to config issues.

Reviewer: ```suggestion
**<ins>Please note</ins>**: an invalid config will cause the Collector to error out and stay down until it is manually restarted using some other mechanism (since the OpAMP extension won't be active). Regular healthchecks and using the Collector's `validate` subcommand to validate config before applying it are **highly recommended** when using this functionality to prevent data loss due to config issues.
```
